### PR TITLE
DirectX bindings

### PR DESCRIPTION
**Pull request by Perksey**
# Summary of the PR
Adds bindings to various DirectX/Windows APIs.

# Related issues, Discord discussions, or proposals
#48 

# Further Comments
Because the JVM was being a big meanie.

- [ ] Direct2D 1
- [x] DXC and D3DCompiler
- [x] Direct3D 9
- [x] Direct3D 11
- [x] Direct3D 12
- [x] DXGI
- [x] XAudio
- [x] XInput
-----
- [x] guids
- [ ] XXXX on 12
-----
- [ ] Metal
- [ ] MetalKit
- [ ] UIKit
- [ ] NS

### CODE REVIEW

**Comment by Perksey** [comment]
- remove pinning in interface methods, replace with unsafe.aspointer
- parse \_In\_ and friends

File: src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs [hunk 1]
```diff
@@ -18,9 +18,9 @@
 namespace Silk.NET.Assimp
 {
     [NativeName("Name", "aiAABB")]
-    public unsafe partial struct AABB
+    public unsafe partial struct Aabb
```
Comment (HurricanKai) [inline src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs:21]: idk what you did to name rules, but that's certainly pretty questionable

File: src/Core/Silk.NET.Core/Native/ComObject.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+// 
+// You may modify and distribute Silk.NET under the terms
+// of the MIT license. See the LICENSE file for details.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Silk.NET.Core.Native
+{
+    public unsafe class ComObject : IDisposable
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/ComObject.cs:11]: explain to me why this has to be in Core please?

File: src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+using System;
+using Silk.NET.Core.Attributes;
+
+#pragma warning disable 1591
+
+namespace Silk.NET.Core.Native
+{
+    [NativeName("Name", "_D3D_CBUFFER_TYPE")]
+    public enum D3DCBufferType
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs:15]: now this really doesn't feel like a core type. it even starts with D3D

File: src/Core/Silk.NET.Core/Native/SilkMarshal.cs [hunk 4]
```diff
@@ -451,6 +462,7 @@ public static GlobalMemory PtrToMemory(IntPtr ptr, int length)
         /// </remarks>
         /// <param name="memory">The memory to zero-initialize the start of.</param>
         /// <param name="memoryLength">The length of the memory.</param>
+        [Obsolete("Random method added in the 2.0 preview cycle, to be removed any day now.")]
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/SilkMarshal.cs:465]: just remove it imo. It's already removed from SilkTouch

File: src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                fixed (void** ppvObjectPtr = &ppvObject)
+                {
+                    return ((delegate* unmanaged[Cdecl]<IUnknown*, Guid*, void**, int>) LpVtbl[0])
+                        (@this, riidPtr, ppvObjectPtr);
+                }
+            }
+        }
+
+        /// <summary>To be added.</summary>
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs:71]: idk why you've started to do this, but don't. Either add something like "To be documented." or just have the warning. "To be added." confuses users.

File: src/Core/Silk.NET.Core/Native/Structs/WindowsMaths.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        public long Width, Height;
+    }
+
+    public struct TagPaletteEntry
+    {
+        public byte PeRed;
+        public byte PeGreen;
+        public byte PeBlue;
+        public byte PeFlags;
+    }
```
Comment (HurricanKai) [inline src/Core/Silk.NET.Core/Native/Structs/WindowsMaths.cs:29]: Remember to replace with `Rectangle<long>`, `Vector2<long>`, `Vector2<long>` and `Vector4<byte>`

File: src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        public override string Linux => "libDXGI.so";
+
+        /// <inheritdoc />
+        public override string MacOS => "libDXGI.dylib";
+
+        /// <inheritdoc />
+        public override string Android => "libDXGI.so";
+
+        /// <inheritdoc />
+        public override string IOS => "__Internal";
```
Comment (HurricanKai) [inline src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs:25]: consider adding a comment indicating that this relies on the override.

**Review by HurricanKai — CHANGES_REQUESTED** [review]
Imo Core should generate as little stuff specific to one thing as possible, and no generated code at all. Not sure why all these D3D types are in Core now.

Also , I can't be bothered to review the 1400 generated files.

File: src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs (hunk 3 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs:15]: Vulkan and OpenGL stuff are in Core. There has never been any guidelines for what constitutes a core addition, just whatever is common across some assemblies.

File: src/Core/Silk.NET.Core/Native/ComObject.cs (hunk 2 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/ComObject.cs:11]: This preps Silk.NET.Core for any kind of COM interop. There has never been any guidelines for what constitutes a core addition, just whatever is common across some assemblies.

File: src/Core/Silk.NET.Core/Native/SilkMarshal.cs (hunk 4 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/SilkMarshal.cs:465]: That PR is still pending, so not yet.

File: src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs (hunk 5 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs:71]: This comment was chosen to remain consistent with the rest of Silk.NET.

File: src/Core/Silk.NET.Core/Native/Structs/WindowsMaths.cs (hunk 6 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/Structs/WindowsMaths.cs:29]: Will do.

File: src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs (hunk 3 above)
Comment (HurricanKai) [reply src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs:15]: I mean yes, perfectly fair point, though I think I've brought up that this is imo a problem too.
I wouldn't complain if it was one or two types, but these are like 10 generated types. Feels like it should just be a separate DX shared lib

File: src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs (hunk 5 above)
Comment (HurricanKai) [reply src/Core/Silk.NET.Core/Native/Structs/IUnknown.cs:71]: fair. will put up a PR to change this everywhere after this has been merged.

File: src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs (hunk 7 above)
Comment (Perksey) [reply src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs:25]: Name containers are not intended for general consumption.

Comment (HurricanKai) [reply src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs:25]: I just mean in regards to code readability 😄 
```suggestion
        public override string IOS => "__Internal"; // __Internal relies on a SilkTouch override.
```
am fine if you don't want this too, just figured I'd nit it

File: src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs (hunk 1 above)
Comment (Perksey) [reply src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs:21]: This makes the bindings consistent with the rest of Silk.NET (i.e. `Sdl`, `Glfw`, etc). The naming changes were required to give some DirectX types nice names. Moot anyway, will be replaced with Box3 when maths is in.

Comment (HurricanKai) [reply src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs:21]: Ok. if it isn't too much of a difficulty make the file name consistent please.

Comment (Perksey) [reply src/Assimp/Silk.NET.Assimp/Structs/AABB.gen.cs:21]: It's pretty difficult given Windows ignores filenames, would require manual git index changes.

File: src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs (hunk 3 above)
Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/Enums/D3DCBufferType.gen.cs:15]: Consumers just don't want to have lots of assemblies. Again, more than happy to discuss a Silk.NET.Core.Slim for non-Silk.NET uses.

File: src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs (hunk 7 above)
Comment (Perksey) [reply src/Microsoft/Silk.NET.DXGI/DXGILibraryNameContainer.cs]: Committed. It's worth noting that there are tens of other name containers.

File: src/Core/Silk.NET.Core/Native/SilkMarshal.cs (hunk 4 above)
Comment (HurricanKai) [reply src/Core/Silk.NET.Core/Native/SilkMarshal.cs:465]: PR updated, maybe we can get that in before this?

Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/SilkMarshal.cs]: Done, will just let CI run through.

Comment (Perksey) [reply src/Core/Silk.NET.Core/Native/SilkMarshal.cs]: CI doesn't indicate any problem.

**Review by HurricanKai — APPROVED** [review]

**Pull request merged**